CI services chapter (GitHub actions) (#1588)
alan-turing-institute/the-turing-way: book/website/reproducible-research/ci/ci-github-actions.md

(rr-ci-ghactions)=
# Continuous Integration with GitHub Actions

> This section will walk you through the basic setup of continuous integration (CI) using **GitHub Actions (GHA)**. GHA is a task automation system fully integrated with GitHub. In other words, it is an API that orchestrates any workflow based on any event. Although there are many CI service providers, GHA makes it easier than ever to incorporate CI into your repositories. It provides a flexible way to automate nearly every aspect of your project workflow. Here are just a few examples of use cases of GitHub Actions:

Reviewer: maybe link to othe main github actions documentation again from the first mention of github actions

Reviewer: Does API defined elsewhere? if not maybe rephrase that sentence

- Automated testing of the software
- Generate reports of the status of any changes in the repository

> - Responding to workflow triggers using labels, issues, `@` mentions, and more

Reviewer: ```suggestion
- Responding to workflow triggers using labels, issues, special mentions, and more
```

- Triggering code reviews and pull requests
- Managing branches

> GitHub Actions are event-driven, which means it responds to any event (`PR` created, issue created, ...) and triggering an action (label, test, sort, etc...). Any collection of these actions is called a workflow. A more detailed description of this GitHub-related Vocabulary is described in the next section.

Reviewer: ```suggestion
GitHub Actions are event-driven, which means it responds to any event (pull request (PR) created, issue created, ...) and triggering an action (label, test, sort, etc...). Any collection of these actions is called a workflow. A more detailed description of this GitHub-related Vocabulary is described in the next section.
```

---
name: ContinuousIntegration-Nov20
alt: An illustration of how continuous integration works with multiple jobs and actions working alongside each other to feed into an illustration of steps to show merging into the main version.
---
On the left: _The Turing Way_ project illustration by Scriberia. Original version on Zenodo. http://doi.org/10.5281/zenodo.3695300. On the right: Overview of the most important concepts of GitHub Actions, adapated from [morioh.com](https://morioh.com/p/aadcfe6cac57).
```

#### 2. Job

**A job** is a section of the workflow, and is made up of one or more steps. In this section of our workflow, the template defines the steps that make up the build job.

#### 3. Step

A **step** represents one effect of the automation. A step could be defined as a GitHub Action, or another unit, like printing something to the console.

#### 4. Actions

A GitHub **Action** is a piece of automation written in a way that is compatible with workflows. Actions can be written by [GitHub](https://github.com/actions), by the open source [community](https://github.com/sdras/awesome-actions), or you can write them yourself!

## Getting started with GitHub Action

> GitHub Actions uses YAML syntax and stored in a directory called `.github/workflows` in the repository. You can either use a templated workflow or create your own.

Reviewer: ```suggestion
GitHub Actions uses YAML syntax for their workflow configuration files which are stored in a directory called `.github/workflows` in the repository. You can either use a templated workflow or create your own.
```

#### 1- Using GitHub Actions template

If you want to get started with GitHub Actions, you can start by clicking the "Actions" tab in the repository where you want to create a workflow, as shown below. Under the "Actions" tab, you will find popular CI workflows, which can help deploy or automate some tasks in the repository.

> ```{figure} ../../figures/gifs/start_ghactions.gif

Reviewer: I love this gif so much, super useful!
